src/libstd/keyword_docs.rs

```diff
@@ -1167,12 +1188,45 @@ mod await_keyword {}
 
 #[doc(keyword = "dyn")]
 //
-/// Name the type of a [trait object].
+/// `dyn` is a prefix of a [trait object]'s type.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The `dyn` keyword is used to highlight that calls to methods on the associated `Trait`
+/// are dynamically dispatched. To use the trait this way, it must be 'object safe'.
+///
+/// Unlike generic parameters or `impl Trait`, the compiler does not know the concrete type that
+/// is being passed. That is, the type has been [erased].
+/// As such, a `dyn Trait` reference contains _two_ pointers.
+/// One pointer goes to the data (e.g., an instance of a struct).
+/// Another pointer goes to a map of method call names to function pointers
+/// (known as a virtual method table or vtable).
+///
+/// At run-time, when a method needs to be called on the `dyn Trait`, the vtable is consulted to get
+/// the function pointer and then that function pointer is called.
+///
+/// ## Trade-offs
+///
+/// The above indirection is the additional runtime cost of calling a function on a `dyn Trait`.
+/// Methods called by dynamic dispatch generally cannot be inlined by the compiler.
+///
+/// However, `dyn Trait` is likely to produce smaller code than `impl Trait` / generic parameters as
+/// the method won't be duplicated for each concrete type.
+///
+/// Read more about `object safety` and [trait object]s.
+///
+/// ## impl Trait for dyn
+///
+/// There is one other location where the `dyn` keyword can be used:
+/// ```rust
+/// # trait TraitA {}
+/// # trait TraitB {}
+/// impl TraitA for dyn TraitB {}
+/// ```
+/// A trait may be implemented for another trait. The implementation is only accessible
+/// through a trait object. While this is limiting, the impl is not bound by object safety
+/// rules and thus can contain `impl Trait` / generic parameters.
```

Comment:
This is in fact *not* another location. This arises purely from the fact that you have:
```rust
impl TraitA for Type { ... }
```
combined with `dyn TraitB` being a valid `Type`. This isn't different than:
```rust
trait A {}
trait B {}
impl A for Box<dyn B> {}
```

The compiler also does not trait `dyn TraitB` specially here other than for coherence.

If you feel this deserves special mention, then I would emphasize how `for dyn B` arises (and avoid saying that "a trait may be implemented for another trait") and how it is *not* a special case.

Reply:
Glad I ran it past you :-)
Ok let me retract and see if we can get this in a shippable state.